gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-client/src/main/java/org/apache/gobblin/util/limiter/BatchedPermitsRequester.java

```diff
@@ -143,21 +149,30 @@ public boolean getPermits(long permits) throws InterruptedException {
     if (permits <= 0) {
       return true;
     }
+    long startTimeNanos = System.nanoTime();
     this.permitsOutstanding.addEntryWithWeight(permits);
     this.lock.lock();
     try {
       while (true) {
+        if (permits > this.knownUnsatisfiablePermits) {
+          // We are requesting more permits than the remote policy will ever be able to satisfy, return immediately with no permits
+          break;
+        }
+        if (elapsedMillis(startTimeNanos) > this.maxTimeout) {
+          // Max timeout reached, break
+          break;
+        }
         if (this.permitBatchContainer.tryTake(permits)) {
           this.permitsOutstanding.removeEntryWithWeight(permits);
           return true;
         }
-        if (this.retryStatus.canRetryWithinMillis(10000)) {
+        if (this.retryStatus.canRetryWithinMillis(remainingTime(startTimeNanos, this.maxTimeout))) {
           long callbackCounterSnap = this.callbackCounter.get();
           maybeSendNewPermitRequest();
           if (this.callbackCounter.get() == callbackCounterSnap) {
             // If a callback has happened since we tried to send the new permit request, don't await
             // Since some request senders may be synchronous, we would have missed the notification
-            this.newPermitsAvailable.await();
+            boolean ignore = this.newPermitsAvailable.await(remainingTime(startTimeNanos, this.maxTimeout), TimeUnit.MILLISECONDS);
```

Comment:
Should this `ignore` variable being used in line 184 to determine if there's indeed necessity to remove permits from `permitsOutstanding?`

Reply:
The permits should always be removed in line 184, regardless of the reason they were not granted. If not, we will end up with a backlog of requests that will never be removed.